system/security/securesocket/securesocket.cpp

```diff
@@ -1142,7 +1142,147 @@ const char* strtok__(const char* s, const char* d, StringBuffer& tok)
     return s;
 }
 
-class CSecureSocketContext : implements ISecureSocketContext, public CInterface
+using BIO_ptr = std::unique_ptr<BIO, decltype(&::BIO_free)>;
+
+static bool usePrivateKeyPEMBuffer(SSL_CTX *ctx, const char *privkeybuf, int privkeylen=-1)
+{
+    // single RSA key in buffer
+
+    BIO_ptr cbio(BIO_new_mem_buf(privkeybuf, privkeylen), ::BIO_free);
+    if (!cbio)
+        return false;
+
+    RSA *rsa = PEM_read_bio_RSAPrivateKey(cbio.get(), NULL, 0, NULL);
+    if (rsa == NULL)
+        return false;
+
+    if (!SSL_CTX_use_RSAPrivateKey(ctx, rsa))
+        return false;
+
+    return true;
+}
+
+static bool useCertificateChainPEMBuffer(SSL_CTX *ctx, const char *certbuf, int certlen=-1)
+{
+    // this routine based on code originally from:
+    // https://stackoverflow.com/questions/3810058/read-certificate-files-from-memory-instead-of-a-file-using-openssl
+
+    // can have multiple certs in buffer
+
+    BIO_ptr cbio(BIO_new_mem_buf(certbuf, certlen), ::BIO_free);
+    if (!cbio)
+        return false;
+
+    STACK_OF(X509_INFO) *infostk = PEM_X509_INFO_read_bio(cbio.get(), NULL, NULL, NULL);
```

Comment:
perhaps a unique_ptr could be used for infostk as well ?

Reply:
yes, I think it would help the code be more robust in general if we contained these objects, such as they were auto deleted when out of scope.